lib/Service/PageService.php

```diff
@@ -391,35 +391,54 @@ private function getIndexPageFile(Folder $folder): File {
 	 */
 	public function getPagesFromFolder(int $collectiveId, Folder $folder, string $userId, bool $recurse = false, bool $forceIndex = false): array {
 		$pageInfos = [];
-		$indexPage = null;
-
-		// Add subpages and recurse over subfolders
-		foreach ($folder->getDirectoryListing() as $node) {
-			if ($node instanceof File && NodeHelper::isPage($node)) {
-				try {
-					$page = $this->getPageByFile($node, $folder);
-				} catch (NotFoundException) {
-					// If parent folder doesn't have an index page, it throws NotFoundException. Let's ignore it.
-					continue;
-				}
-				if (NodeHelper::isIndexPage($node)) {
-					$indexPage = $page;
-				} else {
-					$pageInfos[] = $page;
-				}
-			} elseif ($recurse && $node instanceof Folder) {
-				array_push($pageInfos, ...$this->getPagesFromFolder($collectiveId, $node, $userId, true));
+		$folderNodes = $folder->getDirectoryListing();
+
+		// Get page infos of subfolders
+		if ($recurse) {
+			$subfolders = array_filter($folderNodes, static function (Node $node) {
+				return $node instanceof Folder;
+			});
+			foreach ($subfolders as $subfolder) {
+				array_push($pageInfos, ...$this->getPagesFromFolder($collectiveId, $subfolder, $userId, true));
 			}
 		}
 
-		if (!$indexPage) {
-			if (!$forceIndex && count($pageInfos) === 0) {
+		$pageFiles = array_filter($folderNodes, static function (Node $node) {
+			return $node instanceof File && NodeHelper::isPage($node);
+		});
+
+		// Create missing index page if folder or subfolders have page files (or forceIndex)
+		$hasIndexFile = (bool)array_filter($folderNodes, static function (Node $node) {
+			return $node instanceof File && NodeHelper::isIndexPage($node);
+		});
```

Comment:
Okay, up to three times. bool-casting an array? array_reduce would be more straight forward, but this can be embedded in the same single foreach, no?